### PR TITLE
Make it so this crate can be `no_std`

### PR DESCRIPTION
**Pull request by notgull**
This PR adds an `std` feature enabled on this crate by default. This feature implements `(Ref)UnwindSafe` on the channel and uses `std::thread::yield_now()` in busy-waiting within the channels. Without this feature, `core::hint::spin_loop()` is used instead.

Due to the `(Ref)UnwindSafe` not being implemented without `std`, this is a breaking change.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -29,6 +29,7 @@ jobs:
         if: startsWith(matrix.rust, 'nightly')
         run: cargo check -Z features=dev_dep
       - run: cargo test
+      - run: cargo build --no-default-features
```
Comment (taiki-e) [inline .github/workflows/ci.yml:32]: x86_64-unknown-linux-gnu can link to std, so this does not ensure that no-std is supported.
Also, note that [cargo v1 resolver merges features of dev-dependencies and normal dependency](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification).


To work around these problems, you need a tool like cargo-hack that can work around the v1 resolver problems, and a real no-std target.

```yaml
      - name: Install cargo-hack
        uses: taiki-e/install-action@cargo-hack
      - run: rustup target add thumbv7m-none-eabi
      - run: cargo hack build --target thumbv7m-none-eabi --no-default-features --no-dev-deps


```

**Review by taiki-e — CHANGES_REQUESTED** [review]
Thanks!

File: .github/workflows/ci.yml (hunk 1 above)
Comment (taiki-e) [reply .github/workflows/ci.yml:32]: (Well, we should also fix other no-std crates in smol-rs)

**Review by taiki-e — APPROVED** [review]

**Comment by taiki-e** [comment]
Hmm, the last few commits seem unrelated.

**Comment by notgull** [comment]
Yes, I rebased on master

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        }
+    };
+
+    // We use "no_*" instead of "has_*" here. For non-Cargo
+    // build tools that don't run build.rs, the negative
+    // allows us to treat the current Rust version as the
+    // latest stable version, for when version information
+    // isn't available.
+    if !cfg.probe_rustc_version(1, 49) {
+        autocfg::emit("cq_no_spin_loop");
```
[the remark this answers is not in the document]
Comment (notgull) [reply build.rs]: Whoops, looks like I missed that in the rebase. Should be gone now.

**Review by taiki-e — APPROVED** [review]
Thanks!

**Pull request merged**